### PR TITLE
Extend python higlighter

### PR DESCRIPTION
**Pull request by casimir**
Adds highlight for:
- `self`
- builtin attributes
- builtin methods
- numbers

The symbol lists are based on the online python3's documentation.
The new rules (except for the numbers) can be disabled with an option since it was a concern for certain users, cf seconde commit message.

Some problem remains: 
- [ ] handle weird number formats like `.12`, `1.e2j`

cc @danr

### CODE REVIEW

**Comment by mawww** [comment]
Where are we at with that ? Still WIP ?

**Comment by casimir** [comment]
@mawww there is still some issues with exotic number formats but I guess we can fix that in another PR. Note that this is just for implementation correctness.

File: rc/core/python.kak [hunk 1]
```diff
@@ -47,19 +47,85 @@ add-highlighter -group /python/comment       fill comment
...
+}
+
+# Integer formats
+add-highlighter -group /python/code regex '\b0[bB][01]+[lL]?\b' 0:value
+add-highlighter -group /python/code regex '\b0[xX][\da-fA-F]+[lL]?\b' 0:value
+add-highlighter -group /python/code regex '\b0[oO]?[0-7]+[lL]?\b' 0:value
+add-highlighter -group /python/code regex '\b([1-9]\d*|0)[lL]?\b' 0:value
+%sh{
+    bound_begin='(\h|\v|[({[])'
+    bound_end='(\h|\v|[)}\]])'
```
Comment (danr) [inline rc/core/python.kak]: This seems to be dead code

File: rc/core/python.kak [hunk 2]
```diff
@@ -47,19 +47,85 @@ add-highlighter -group /python/comment       fill comment
...
+    values="self"
+    # curl -s "https://docs.python.org/3/reference/datamodel.html#special-method-names" | ag -o '__[a-z_]+__[^(]' | sed s'/.$//' | sort | uniq
+    attributes="__abs__|__add__|__aenter__|__aexit__|__aiter__|__and__|__anext__"
+    attributes="${attributes}|__annotations__|__await__|__bases__|__bool__|__bytes__"
+    attributes="${attributes}|__call__|__class__|__classcell__|__closure__|__code__"
+    attributes="${attributes}|__complex__|__contains__|__defaults__|__del__"
+    attributes="${attributes}|__delattr__|__delete__|__delitem__|__dict__|__dir__"
+    attributes="${attributes}|__divmod__|__doc__|__enter__|__eq__|__exit__|__file__"
+    attributes="${attributes}|__float__|__floordiv__|__format__|__func__|__future__"
+    attributes="${attributes}|__ge__|__get__|__getattr__|__getattribute__"
```
Comment (danr) [inline rc/core/python.kak]: I don't understand the distinction between what you have called `attributes` and  what you have called `methods`. Some of them are listed in both for instance `__ge__` here...

File: rc/core/python.kak [hunk 3]
```diff
@@ -47,19 +47,85 @@ add-highlighter -group /python/comment       fill comment
...
+    attributes="${attributes}|__rfloordiv__|__rlshift__|__rmatmul__|__rmod__"
+    attributes="${attributes}|__rmul__|__ror__|__round__|__rpow__|__rrshift__"
+    attributes="${attributes}|__rshift__|__rsub__|__rtruediv__|__rxor__|__self__"
+    attributes="${attributes}|__set__|__setattr__|__setitem__|__set_name__"
+    attributes="${attributes}|__slots__|__str__|__sub__|__subclasscheck__"
+    attributes="${attributes}|__truediv__|__weakref__|__xor__"
+    # curl -s "https://docs.python.org/3/reference/datamodel.html#special-method-names" | ag -o '__[a-z_]+__\| | sort | uniq
+    methods="__add__|__anext__|__await__|__bool__|__call__|__contains__|__del__"
+    methods="${methods}|__delattr__|__delete__|__delitem__|__divmod__|__enter__"
+    methods="${methods}|__eq__|__exit__|__floordiv__|__format__|__ge__|__get__"
```
Comment (danr) [inline rc/core/python.kak]: ... and `__ge__` here.

The argument to `ag` seems incomplete, `'` is never matched (?)

**Comment by casimir** [comment]
Ok so I just figured out I pushed a wrong version of the script. @danr I how you an apology for making you review an incomplete PR!
Anyway I just rebased the PR with the correct version. It can still be incomplete but I use this at work for a solid 2 months and I'm mostly happy with it.

File: rc/core/python.kak [hunk 4]
```diff
@@ -47,19 +50,64 @@ add-highlighter -group /python/comment       fill comment
...
     printf %s "
-        add-highlighter -group /python/code regex '\b(${types})\b' 0:type
+        add-highlighter -group /python/code regex '\b(${types})\b' 1:type
         add-highlighter -group /python/code regex '@[\w_]+\b' 0:attribute
     "
 }
 
+# extended syntax
+
+%sh{ if [ "$kak_opt_python_use_extended_syntax" = "true" ]; then
```
Comment (lenormf) [inline rc/core/python.kak]: I understand the logic behind this, but at this point if there isn't going to be a separate file for extended highlighting I think it's better if everything is merged at once. Having options to control the behaviour of a script at load time really isn't part of the editor's paradigm, as behaviour is mostly controlled by what script is loaded or not.

File: rc/core/python.kak [hunk 5]
```diff
@@ -47,19 +50,64 @@ add-highlighter -group /python/comment       fill comment
...
+    methods="${methods}|__reversed__|__rfloordiv__|__rlshift__|__rmatmul__|__rmod__"
+    methods="${methods}|__rmul__|__ror__|__round__|__rpow__|__rrshift__|__rshift__"
+    methods="${methods}|__rsub__|__rtruediv__|__rxor__|__set__|__setattr__"
+    methods="${methods}|__setitem__|__set_name__|__slots__|__str__|__sub__"
+    methods="${methods}|__truediv__|__xor__"
+
+    printf %s "
+        add-highlighter -group /python/code regex '\b(${values})\b' 1:variable
+        add-highlighter -group /python/code regex '\.(${attributes})[^(]' 1:attribute
+        add-highlighter -group /python/code regex '(def\s+|\.)(${methods})\(' 2:function
```
Comment (lenormf) [inline rc/core/python.kak]: I think `\bdef` would be better.

File: rc/core/python.kak [hunk 6]
```diff
@@ -47,19 +50,64 @@ add-highlighter -group /python/comment       fill comment
...
+add-highlighter -group /python/code regex '\b0[bB][01]+[lL]?\b' 0:value
+add-highlighter -group /python/code regex '\b0[xX][\da-fA-F]+[lL]?\b' 0:value
+add-highlighter -group /python/code regex '\b0[oO]?[0-7]+[lL]?\b' 0:value
+add-highlighter -group /python/code regex '\b([1-9]\d*|0)[lL]?\b' 0:value
+# Float formats
+add-highlighter -group /python/code regex '(\h|\v|[({[])(-((\d+\.\d+)|(\d+\.)|(\.\d+)))([eE][+-]?\d+)?(\h|\v|[)}\]])' 2:value
+add-highlighter -group /python/code regex '\b\d+[eE][+-]?\d+\b' 0:value
+# Imaginary formats
+add-highlighter -group /python/code regex '\b\d+[jJ]\b' 0:value
+add-highlighter -group /python/code regex '(\h|\v|[({[])(-((\d+\.\d+)|(\d+\.)|(\.\d+)))([eE][+-]?\d+)?[jJ](\h|\v|[)}\]])' 2:value
```
Comment (lenormf) [inline rc/core/python.kak]: Have you tested this regex against a list of huge imaginary numbers? Would adding non matching patterns to useless groups help?

**Comment by lenormf** [comment]
#1699 has been merged and superseeds this PR, can it be closed?

**Comment by casimir** [comment]
👍

**Pull request closed without merging**